contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -927,6 +928,208 @@ checksum(PG_FUNCTION_ARGS)
         PG_RETURN_INT32(result);
 }
 
+Datum
+object_id(PG_FUNCTION_ARGS)
+{	
+	char 			*db_name = "";
+	char 			*schema_name = "";
+	char 			*physical_schema_name;
+	char 			*object_name;
+	bool 			is_temp_object;
+	Oid 			schema_oid;
+	Oid				user_id = GetUserId();
+	Oid 			result = InvalidOid;
+	int 			i;
+	char 			*input = lowerstr(text_to_cstring(PG_GETARG_TEXT_P(0)));
+	char 			*object_type = text_to_cstring(PG_GETARG_TEXT_P(1));
+	char 			**splited_object_name;
+
+	/* strip trailing whitespace */
+	i = strlen(input);
+	while (i > 0 && isspace((unsigned char) input[i - 1]))
+		input[--i] = '\0';
+	
+	/* length should be restricted to 4000 */
+	if(strlen(input) > 4000)
+		ereport(ERROR,
+				(errcode(ERRCODE_STRING_DATA_RIGHT_TRUNCATION),
+				errmsg("String or binary data would be truncated.")));
+
+	/* resolve the three part name */
+	splited_object_name = split_object_name(input);
+	db_name = splited_object_name[1];
+	schema_name = splited_object_name[2];
+	object_name = splited_object_name[3];
+
+	pfree(input);
+	pfree(splited_object_name);
+
+	/* truncate identifiers if needed */
+	truncate_tsql_identifier(db_name);
+	truncate_tsql_identifier(schema_name);
+	truncate_tsql_identifier(object_name);
+
+	if (!strcmp(db_name, ""))
+		db_name = get_cur_db_name();
+	else if (strcmp(db_name, get_cur_db_name()) && strcmp(db_name, "tempdb"))
+	{
+		/* cross database lookup */
+		const char *login;
+		const char *user = NULL;
+		int db_id = get_db_id(db_name);
+
+		if (!DbidIsValid(db_id))
+			PG_RETURN_NULL();
+		login = GetUserNameFromId(GetSessionUserId(), false);
+		user = get_authid_user_ext_physical_name(db_name, login);
+		if(!user)
+			PG_RETURN_NULL();
```

Comment:
Why do we need this logic?
Because it will return NULL(because won't have any entry in user_ext catalog) for 2 valid cases:
1. When login is member of sysadmin, in this case user will be dbo.
2. When that database has guest user enabled, in that case it will by default use guest.

For more info. please refer `Datum has_dbaccess(PG_FUNCTION_ARGS)` which contains logic around above 2 cases.

We can just use loginid in case of cross-db, it will handle all the cases.